includes/REST_API/Link_Controller.php

```diff
@@ -0,0 +1,255 @@
+ */
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\REST_API;
+
+use DOMDocument;
+use DOMNodeList;
+use DOMXpath;
+use WP_Error;
+use WP_Http;
+use WP_REST_Controller;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * API endpoint to allow parsing of metadata from a URL
+ *
+ * Class Link_Controller
+ */
+class Link_Controller extends WP_REST_Controller {
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'link';
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'parse_link' ],
+					'permission_callback' => [ $this, 'parse_link_permissions_check' ],
+					'args'                => [
+						'url'       => [
+							'description'       => __( 'The URL to process.', 'web-stories' ),
+							'type'              => 'string',
+							'required'          => true,
+							'sanitize_callback' => 'esc_url_raw',
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Parses a URL to return some metadata for inserting links.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function parse_link( $request ) {
+		$url = $request['url'];
+
+		/**
+		 * Filters the link data TTL value.
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 * @param string $url The attempted URL.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_link_data_cache_ttl', DAY_IN_SECONDS, $url );
+		$cache_key = 'web_stories_link_data_' . md5( $url );
+
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$title       = '';
+		$image       = '';
+		$description = '';
+
+		$data = [
+			'title'       => $title,
+			'image'       => $image,
+			'description' => $description,
+		];
+
+		$response = wp_safe_remote_get(
+			$url,
+			[
+				'limit_response_size' => 153600, // 150 KB.
+			]
+		);
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
```

Comment:
If `$response` is a `WP_Errror` object here, should we just return it. Might be helpful in debugging issues.

Reply:
I hesitate a bit to just forward these errors. It also depends a bit on how these would be surfaced in the UI.

I'd suggest leaving this for now and adjust later if necessary.